### DISCUSSION TITLE
SD2.1 512 NonEMA VS SD2.1 768 NonEMA - Memory Usage

### DISCUSSION THREAD

**Discussion by mediocreatmybest** [d8ahazard/sd_dreambooth_extension#619] [Q&A] [answered]
Hi, everyone.

Something I've noticed when trying to train SD2.1 (all unsuccessfully with 12GB vRAM on Linux).

A) When creating the model from the check point SD_Dreambooth_extension assumes all versions of SDv2.x are using "prediction_type": "v_prediction" with the scheduler. Looking at the original yaml file, V prediction doesn't look like its part of the base 2.1 512 model. (Something I've mentioned in a previous post) So I'm unsure exactly if this causes errors or issues with training? 

B) When attempting to train SDv2.1 512 NonEMA Base, memory usage is over 10GB and as a consequence the training fails due to lack of memory. 

C) When attempting to train SDv2.1 768 NonEMA, Memory usage is quite low at approx  5GB. I suspect something is causing an error and not running correctly as the loss (as mentioned in another post) instantly reports NaN. 

All of this is with the latest Automatic1111 and the latest update of the SD_dreambooth_extension (at time of post) 
The model is flagged as half model, fp16, 8bitAdam, Flash_Attention. 

Note: I'm using Flash_Attention due to Xformers throwing up CUDA "Invalid argument"
I'm happy to post an issue with this, but honestly it is hard to know where the glitch is.  

Problem with Xformers? Problem with Flash_Attention? Problem with SD2.x training in general?  

I can successfully train SD1.5 as finetuning with Flash_attention without issues at all. 

I have CUDA11.7 / Python 3.10.6 / Xformers compiled manually as no wheels will install. 

I've also listed my Python version numbers in the venv, so if anyone sees any glaring version errors. 

Cheers,

```
stable-diffusion-webui$ pip list
Package                      Version
---------------------------- ------------------------
absl-py                      1.3.0
accelerate                   0.15.0
addict                       2.4.0
aenum                        3.1.11
aiohttp                      3.8.3
aiosignal                    1.3.1
albumentations               1.3.0
antlr4-python3-runtime       4.9.3
anyio                        3.6.2
asttokens                    2.2.1
astunparse                   1.6.3
async-timeout                4.0.2
attrs                        22.2.0
backcall                     0.2.0
basicsr                      1.4.2
bcrypt                       4.0.1
beautifulsoup4               4.11.1
bitsandbytes                 0.35.0
blendmodes                   2022
boltons                      21.0.0
cachetools                   5.2.0
certifi                      2022.12.7
cffi                         1.15.1
chardet                      4.0.0
charset-normalizer           2.1.1
clean-fid                    0.1.29
click                        8.1.3
clip                         1.0
cmake                        3.25.0
coloredlogs                  15.0.1
contourpy                    1.0.6
cryptography                 38.0.4
cuda-python                  11.7.0
cycler                       0.11.0
Cython                       0.29.32
decorator                    5.1.1
deepspeed                    0.7.7
deprecation                  2.1.0
diffusers                    0.10.2
einops                       0.4.1
executing                    1.2.0
facexlib                     0.2.5
fairscale                    0.4.9
fastapi                      0.87.0
ffmpy                        0.3.0
filelock                     3.8.2
filterpy                     1.4.5
flash-attn                   0.2.5
flatbuffers                  22.12.6
font-roboto                  0.0.1
fonts                        0.0.3
fonttools                    4.38.0
frozenlist                   1.3.3
fsspec                       2022.11.0
ftfy                         6.1.1
future                       0.18.2
gast                         0.4.0
gdown                        4.6.0
gfpgan                       1.3.8
gitdb                        4.0.10
GitPython                    3.1.27
google-auth                  2.15.0
google-auth-oauthlib         0.4.6
google-pasta                 0.2.0
gradio                       3.9
grpcio                       1.51.1
h11                          0.12.0
h5py                         3.7.0
hjson                        3.1.0
httpcore                     0.15.0
httpx                        0.23.1
huggingface-hub              0.11.1
humanfriendly                10.0
idna                         2.10
imageio                      2.23.0
importlib-metadata           5.2.0
inflection                   0.5.1
ipython                      8.6.0
jedi                         0.18.2
Jinja2                       3.1.2
joblib                       1.2.0
jsonmerge                    1.8.0
jsonschema                   4.17.3
keras                        2.11.0
kiwisolver                   1.4.4
kornia                       0.6.7
lark                         1.1.2
libclang                     14.0.6
linkify-it-py                1.0.3
llvmlite                     0.39.1
lmdb                         1.4.0
lpips                        0.1.4
Markdown                     3.4.1
markdown-it-py               2.1.0
MarkupSafe                   2.1.1
matplotlib                   3.6.2
matplotlib-inline            0.1.6
mdit-py-plugins              0.3.3
mdurl                        0.1.2
modelcards                   0.1.6
mpmath                       1.2.1
multidict                    6.0.3
mypy-extensions              0.4.3
networkx                     2.8.8
ninja                        1.11.1
numba                        0.56.4
numpy                        1.23.3
nvidia-pyindex               1.0.9
oauthlib                     3.2.2
omegaconf                    2.2.3
onnxruntime-gpu              1.13.1
open-clip-torch              2.6.1
opencv-python                4.6.0.66
opencv-python-headless       4.6.0.66
opt-einsum                   3.3.0
orjson                       3.8.3
packaging                    22.0
pandas                       1.5.2
paramiko                     2.12.0
parso                        0.8.3
pexpect                      4.8.0
pickleshare                  0.7.5
piexif                       1.1.3
Pillow                       9.2.0
pip                          22.3.1
prompt-toolkit               3.0.36
protobuf                     3.19.6
psutil                       5.9.4
ptyprocess                   0.7.0
pure-eval                    0.2.2
py-cpuinfo                   9.0.0
pyasn1                       0.4.8
pyasn1-modules               0.2.8
pycparser                    2.21
pycryptodome                 3.16.0
pydantic                     1.10.2
pyDeprecate                  0.3.2
pydub                        0.25.1
Pygments                     2.13.0
PyNaCl                       1.5.0
pyparsing                    3.0.9
pyre-extensions              0.0.23
pyrsistent                   0.19.2
PySocks                      1.7.1
python-dateutil              2.8.2
python-multipart             0.0.4
pytorch-lightning            1.7.6
pytz                         2022.7
PyWavelets                   1.4.1
PyYAML                       6.0
qudida                       0.0.4
realesrgan                   0.3.0
regex                        2022.10.31
requests                     2.25.1
requests-oauthlib            1.3.1
resize-right                 0.0.2
rfc3986                      1.5.0
rsa                          4.9
safetensors                  0.2.5
scikit-image                 0.19.2
scikit-learn                 1.2.0
scipy                        1.9.3
seaborn                      0.12.1
sentencepiece                0.1.97
setuptools                   65.6.3
six                          1.16.0
smmap                        5.0.0
sniffio                      1.3.0
soupsieve                    2.3.2.post1
stack-data                   0.6.2
starlette                    0.21.0
sympy                        1.11.1
tb-nightly                   2.12.0a20221220
tensorboard                  2.11.0
tensorboard-data-server      0.6.1
tensorboard-plugin-wit       1.8.1
tensorflow                   2.11.0
tensorflow-cpu               2.11.0
tensorflow-estimator         2.11.0
tensorflow-io-gcs-filesystem 0.29.0
termcolor                    2.1.1
threadpoolctl                3.1.0
tifffile                     2022.10.10
timm                         0.6.7
tokenizers                   0.12.1
torch                        1.12.1+cu113
torchdiffeq                  0.2.3
torchmetrics                 0.11.0
torchsde                     0.2.5
torchvision                  0.13.1+cu113
tqdm                         4.64.1
traitlets                    5.8.0
trampoline                   0.1.2
transformers                 4.25.1
triton                       2.0.0.dev20221202
typing_extensions            4.4.0
typing-inspect               0.8.0
uc-micro-py                  1.0.1
urllib3                      1.26.13
uvicorn                      0.20.0
wcwidth                      0.2.5
websockets                   10.4
Werkzeug                     2.2.2
wheel                        0.38.4
wrapt                        1.14.1
xformers                     0.0.15+7e05e2c.d20221223
yapf                         0.32.0
yarl                         1.8.2
zipp                         3.11.0
```

**Comment by mediocreatmybest** [comment]
After a quick look around for SDv2.1 512 Base prediction type. It looks like the prediction is "epsilon" the same as the previous SDv1.x, where SDv2.x 768 looks to be v_prediction. I'm guessing this may cause issues with the conversion to and from diffusers to a checkpoint?

**Comment by mediocreatmybest** [comment, marked as answer]
https://huggingface.co/stabilityai/stable-diffusion-2-1-base/blob/main/scheduler/scheduler_config.json

For the SDv2.1 Base 512. 

```
{
  "_class_name": "PNDMScheduler",
  "_diffusers_version": "0.10.0.dev0",
  "beta_end": 0.012,
  "beta_schedule": "scaled_linear",
  "beta_start": 0.00085,
  "clip_sample": false,
  "num_train_timesteps": 1000,
  "prediction_type": "epsilon",
  "set_alpha_to_one": false,
  "skip_prk_steps": true,
  "steps_offset": 1,
  "trained_betas": null
}

```

**Reply by mediocreatmybest** [reply]
Ok, not sure if this is related, but I'm suspecting this is part of the issue with SD2.1. I haven't tried with 2.0 yet to confirm. 

https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/5715

Xformers with 2.1 doesn't seem to be very happy. 

I'm finding I can't train hypernetworks with 2.1 and it seems to have the exact same symptoms.

**Comment by mediocreatmybest** [comment]
- [x] Seems to be all working as intended now.
